### PR TITLE
Add web support to expo-keep-awake

### LINKED ISSUE #6889: [Web] expo-keep-awake support
https://whatwebcando.today/wake-lock.html

### PR DESCRIPTION
**Pull request by EvanBacon**
# Why

- Reimplements https://github.com/expo/expo/pull/9467 which is stale.
- Resolves https://github.com/expo/expo/issues/6889

# How

- `KeepAwake.activateKeepAwake` now returns a promise to accommodate web.
- Added `KeepAwake.isAvailableAsync` which returns false on certain web browsers.
- Added `KeepAwake.addListener` to observe state changes on web.

# Test Plan

# Checklist

- [x] Documentation is up to date to reflect these changes (eg: https://docs.expo.dev and README.md).
- [x] This diff will work correctly for `expo build` (eg: updated `@expo/xdl`).
- [x] This diff will work correctly for `expo prebuild` & EAS Build (eg: updated a module plugin).

### CODE REVIEW

File: packages/expo-keep-awake/src/ExpoKeepAwake.web.ts [hunk 1]
```diff
@@ -1 +1,51 @@
...
+    if (Platform.isDOMAvailable) {
+      return 'wakeLock' in navigator;
+    }
+    return false;
+  },
+  async activate(tag: string) {
+    if (!Platform.isDOMAvailable) {
+      return;
+    }
+    const wakeLock = await (navigator as any).wakeLock.request('screen');
```
Comment (bbarthec) [inline packages/expo-keep-awake/src/ExpoKeepAwake.web.ts:19]: I would rather go with polyfilling the type system with the missing typings. The very same way we do it in other modules:
https://github.com/expo/expo/blob/cd195fb2d1fd09d1d1ea8dad569065690b884a6e/packages/expo-cellular/src/ts-declarations/lib.dom.d.ts#L7-L18
and then no `as any` casting is happening:
https://github.com/expo/expo/blob/cd195fb2d1fd09d1d1ea8dad569065690b884a6e/packages/expo-cellular/src/ExpoCellular.web.ts#L20-L21

**Review by bbarthec — CHANGES_REQUESTED** [review]
Looking awesome 🚀 
I just have a small remark regarding missing typings.

File: packages/expo-keep-awake/src/index.ts [hunk 2]
```diff
@@ -43,3 +57,31 @@ export function activateKeepAwake(tag: string = ExpoKeepAwakeTag): void {
...
+): Subscription {
+  // Assert so the type is non-nullable.
+  if (!ExpoKeepAwake.addListener) {
+    throw new UnavailabilityError('ExpoKeepAwake', 'addListener');
+  }
+
+  const tag = typeof tagOrListener === 'string' ? tagOrListener : ExpoKeepAwakeTag;
+  const _listener = typeof tagOrListener === 'function' ? tagOrListener : listener;
+
+  return ExpoKeepAwake.addListener(tag, 'ExpoNavigationBar.didChange', _listener);
```
Comment (brentvatne) [inline packages/expo-keep-awake/src/index.ts:84]: ExpoNavigationBar?

File: packages/expo-keep-awake/src/index.ts [hunk 3]
```diff
@@ -26,11 +38,13 @@ export function useKeepAwake(tag: string = ExpoKeepAwakeTag): void {
...
  * each one in order to re-enable screen sleep. If tag is unspecified, the default `tag` is used.
+ *
+ * Web support [is limited](https://caniuse.com/wake-lock).
+ *
  * @param tag *Optional* - Tag to lock screen sleep prevention. If not provided, the default tag is used.
  */
-
-export function activateKeepAwake(tag: string = ExpoKeepAwakeTag): void {
-  if (ExpoKeepAwake.activate) ExpoKeepAwake.activate(tag);
+export async function activateKeepAwake(tag: string = ExpoKeepAwakeTag): Promise<void> {
```
Comment (brentvatne) [inline packages/expo-keep-awake/src/index.ts:46]: maybe create new method `activateKeepAwakeAsync` then keep the old one around and add a deprecation warning?

**Review by brentvatne — APPROVED** [review]

**Review by bbarthec — APPROVED** [review]

**Comment by dvkellerman** [comment]
Hello @EvanBacon, @brentvatne !
Do you need some help with this PR?

File: packages/expo-keep-awake/src/KeepAwake.types.ts [hunk 4]
```diff
@@ -11,3 +11,14 @@ export enum KeepAwakeEventState {
...
 export type KeepAwakeListener = (event: KeepAwakeEvent) => void;
+
+export type KeepAwakeOptions = {
+  /**
+   * The call will throw an unhandled promise rejection on Android when the original Activity is dead or deactivated.
+   * Set the value to `true` for suppressing the uncaught exception.
+   */
+  suppressDeactivateWarnings?: boolean;
+
+  /** A callback that is invoked when the keep-awake state changes (web-only). */
```
Comment (Simek) [inline packages/expo-keep-awake/src/KeepAwake.types.ts:22]:
```suggestion
  /** 
   * A callback that is invoked when the keep-awake state changes.
   * @platform web
   */
```

We can use the `@platform` tag here.

**Pull request merged**